### ISSUE TITLE
settings.alert_mask of lt::alert_category_t::all() prevents add_torrent_alert

### ISSUE THREAD

**Issue by ghbplayer** [arvidn/libtorrent#4107]
libtorrent version (or branch): 1.2 from the other day

platform/architecture: ubuntu 19.10 AMD x64

compiler and compiler version: g++ 9.2.1 20191008 c++17

When creating a session, if I set alert_mask to all() (or -1) I never get add_torrent_alert alerts (which is a problem). If I set it to _lt::alert::error_notification | lt::alert::storage_notification | lt::alert::status_notification_, I do get notifications.

Building as:
boost/b2 -sBOOST_ROOT=boost -j16 variant=debug dht=on asserts=on encryption=on crypto=libcrypto link=static invariant-checks=on debug-symbols=on deprecated-functions=off i2p=off profile-calls=on extensions=on mutable-torrents=off

Edit: further testing indicates any of these are enough to trigger the bug:
lt::alert::port_mapping_notification |lt::alert::port_mapping_log_notification|lt::alert::dht_log_notification

e.g.
lt::alert::error_notification | lt::alert::storage_notification | lt::alert::status_notification | **lt::alert::port_mapping_notification**

**Comment by arvidn** [comment]
`add_torrent_alert` has been deprecated for a long time and will only be available if you build libtorrent with deprecated functionality enabled.

I would suggest you use `add_torrent_alert` instead.

**Comment by arvidn** [comment]
I would have expected you to get deprecation warnings for using `add_torrent_alert`, but looking into this it seems I removed them from the `TORRENT_EXPORT_DEPRECATED` macro. I've fixed that here, does it make sense?

https://github.com/arvidn/libtorrent/pull/4109

**Comment by ghbplayer** [comment]
Yes, I'm confused. Use `add_torrent_alert` rather than `add_torrent_alert`?

I see you made a new branch _fix-deprecation-warnings_. I switched to that branch using `git checkout fix-deprecation-warnings` and built using:
`boost/b2 -sBOOST_ROOT=boost -j16 cxxstd=17 -g variant=debug dht=on asserts=on encryption=on crypto=openssl link=static invariant-checks=on debug-symbols=on deprecated-functions=off i2p=off profile-calls=on extensions=on mutable-torrents=off`

I then built and linked my own library against libtorrent without any warnings. I should probably mentioned I'm using `#define TORRENT_ABI_VERSION 2` in my pch, I haven't seen any mention that I should do so but it seemed reasonable.

Still, I'm not sure why we're talking about deprecate. My original use case was to enable all alerts using `all()` and use `session::async_add_torrent()` as described in the documentation and tutorial. Is there another way?

**Comment by arvidn** [comment]
heh, I meant `torrent_added_alert` (I promise, I even looked it up in the source, I just transcribed it wrong)

**Comment by arvidn** [comment]
ok, I see now that I was confused.

**Comment by arvidn** [comment]
see if you get `alerts_dropped_alert`. If you do, and it indicates the `add_torrent_alert` was dropped, you may need to:

1. pop alerts more frequently
2. increase alert queue size

**Comment by arvidn** [comment]
given that you enable all alerts, you'll probably get so many alerts that the default alert queue size is saturated quite quickly.

sorry it took me so long to realize what this ticket was about :/

**Comment by ghbplayer** [comment]
When things are running I can get flooded with alerts, to handle that I've set alert_queue_size to 10000. However, in this scenario, the problem is more the opposite. I just did a test run with almost every alert enabled. It looks more like everything just stalls, even without adding a torrent I stop getting any alerts within 1 second of program startup.

Perhaps I should make an example program to demonstrate the problem.
